bpf/bpf_overlay.c

```diff
@@ -333,6 +339,51 @@ int tail_handle_ipv4(struct __ctx_buff *ctx)
 					      CTX_ACT_DROP, METRIC_INGRESS);
 	return ret;
 }
+
+#ifdef ENABLE_VTEP
+/*
+ * ARP responder for ARP requests from VTEP
+ * Respond to remote VTEP endpoint with cilium_vxlan MAC
+ */
+__section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_ARP)
+int tail_handle_arp(struct __ctx_buff *ctx)
+{
+	union macaddr mac = NODE_MAC;
+	union macaddr smac;
+	__be32 sip;
+	__be32 tip;
+	__u32 monitor = 0;
+	int i;
+	int ret;
+	struct remote_endpoint_info *info;
+	struct bpf_tunnel_key key = {};
+
+	if (unlikely(ctx_get_tunnel_key(ctx, &key, sizeof(key), 0) < 0))
+		return send_drop_notify_error(ctx, 0, DROP_NO_TUNNEL_KEY, CTX_ACT_DROP,
+										METRIC_INGRESS);
+
+	if (!arp_validate(ctx, &mac, &smac, &sip, &tip))
+		return CTX_ACT_OK;
+	if (!__lookup_ip4_endpoint(tip))
+		return CTX_ACT_OK;
+	info = lookup_ip4_remote_endpoint(sip);
+	if (!info)
+		return CTX_ACT_OK;
```

Comment:
I just realized that there's no notifications for these case. When handing traffic to the stack, we usually do `send_trace_notify(ctx, TRACE_TO_STACK, ...)`.

All of the other return conditions already trigger notifications to userspace, so it's just these three `CTX_ACT_OK` cases that would need to be updated to also call `send_trace_notify(...)`. Could you look into that?

Reply:
what trace reason would that be for these `CTX_ACT_OK`?

```
/* Reasons for forwarding a packet. */
enum trace_reason {
        TRACE_REASON_POLICY = CT_NEW,
        TRACE_REASON_CT_ESTABLISHED = CT_ESTABLISHED,
        TRACE_REASON_CT_REPLY = CT_REPLY,
        TRACE_REASON_CT_RELATED = CT_RELATED,
        TRACE_REASON_CT_REOPENED = CT_REOPENED,
        TRACE_REASON_ENCRYPTED = 0x80
} __packed;
```

Reply:
@joestringer  `tail_handle_arp` in `bpf/bpf_lxc.c` did not have `send_trace_notify(...)` for `CTX_ATC_OK`, do we really need it?

Reply:
It looks like for other cases where there's not a great mapping in TRACE_REASON, we just pass `0` for the reason field.

I suspect it's just an oversight in `tail_handle_arp()`, that function calls the older function `arp_respond()` which _does_ send the notifications. While it's not strictly required for connectivity, this is how Hubble can see the packets flowing through the system. Unless I hear otherwise from @cilium/sig-datapath / @cilium/sig-hubble , I think it makes sense to get this kind of visibility into the traffic to assist debugging if anything goes wrong.

Reply:
ok, I am wondering if I can group all the `CTX_ACT_OK` case with `||` condition in one `if`  to reduce the line of codes.

```
if ( !arp_validate(ctx, &mac, &smac, &sip, &tip) || !__lookup_ip4_endpoint(tip) || !lookup_ip4_remote_endpoint(sip)) {
     send_trace_notify(...);
     return CTX_ACT_OK;
}
```

Reply:
Another reasonable alternative is a goto statement to the end of the function where after the current drop, you have a "to_stack:" label which does the trace + return CTX_ACT_OK.

The problem with combining them is mainly for the `lookup_ip4_remote_endpoint()` call where you actually need the return `info` variable for code below.

Reply:
good idea, i used goto, did not save code lines, but I feel it is better :) pushed the changes, please take a look to see if I am missing anything.